### PR TITLE
[cmd/mdatagen] add doc generation capabilities

### PR DESCRIPTION
**Pull request by atoulme**
**Description:**
This adds documentation generation capabilities to mdatagen. Those capabilities are taken after configschema.

**Link to tracking Issue:** 
#30187

**Testing:**
Unit tests, and a visual representation with apachereceiver.

### CODE REVIEW

File: .github/ALLOWLIST [hunk 1]
```diff
@@ -24,3 +24,4 @@ internal/common
 ## DEPRECATED components
 
 ## UNMAINTAINED components
+exporter/alibabacloudlogserviceexporter/
```
Comment (mx-psi) [inline .github/ALLOWLIST:27]: This should go on a separate PR, right?

File: .github/CODEOWNERS [hunk 2]
```diff
@@ -290,3 +290,4 @@ reports/distributions/liatrio.yaml  @open-telemetry/collector-contrib-approvers
 
 ## UNMAINTAINED components
 
+exporter/alibabacloudlogserviceexporter/                 @open-telemetry/collector-contrib-approvers
```
Comment (mx-psi) [inline .github/CODEOWNERS:293]: ditto

File: cmd/mdatagen/loader.go [hunk 3]
```diff
@@ -226,6 +230,8 @@ type metadata struct {
 	ShortFolderName string `mapstructure:"-"`
 
 	Tests *tests `mapstructure:"tests"`
+
+	Config *config `mapstructure:"config"`
```
Comment (mx-psi) [inline cmd/mdatagen/loader.go:234]: I find the 

```
config:
  generate: true
```

confusing.

My suggestion would be something like

```
config_documentation:
   enabled: true
```

or even

```
config:
  documentation:
     enabled: true
```

where it is clear that we are talking about the configuration documentation and not about generating the configuration structs themselves.

**Comment by dmitryax** [comment]
How much different `internal/configschema` from `cmd/configschema`? Is there a chance to reuse the code and make `cmd/configschema` depend on `internal/configschema` or the other way around?

**Comment by atoulme** [comment]
> How much different `internal/configschema` from `cmd/configschema`? Is there a chance to reuse the code and make `cmd/configschema` depend on `internal/configschema` or the other way around?

I thought we'd let `cmd/configschema` rot and remove it cleanly while we iterate on `internal/configschema`.

File: cmd/mdatagen/loader.go (hunk 3 above)
Comment (atoulme) [reply cmd/mdatagen/loader.go:234]: sounds good to me!

File: .github/ALLOWLIST (hunk 1 above)
Comment (atoulme) [reply .github/ALLOWLIST:27]: it probably got added during the regeneration of the CODEOWNERS file.

File: .github/CODEOWNERS (hunk 2 above)
Comment (atoulme) [reply .github/CODEOWNERS]: fixed with a rebase now, thanks!

File: .github/ALLOWLIST (hunk 1 above)
Comment (atoulme) [reply .github/ALLOWLIST]: fixed with a rebase now, thanks!

**Comment by mx-psi** [comment]
> > How much different `internal/configschema` from `cmd/configschema`? Is there a chance to reuse the code and make `cmd/configschema` depend on `internal/configschema` or the other way around?
> 
> I thought we'd let `cmd/configschema` rot and remove it cleanly while we iterate on `internal/configschema`.

If it's not too hard, I would prefer doing what @dmitryax said and reuse `internal/configschema` for `cmd/configschema`. It's the approach we usually have taken when moving code around

File: .github/CODEOWNERS [hunk 4]
```diff
@@ -113,6 +113,7 @@ extension/storage/filestorage/                           @open-telemetry/collect
 
 internal/aws/                                            @open-telemetry/collector-contrib-approvers @Aneurysm9 @mxiamxia
 internal/collectd/                                       @open-telemetry/collector-contrib-approvers @atoulme
+internal/configschema/                                   @open-telemetry/collector-contrib-approvers @mx-psi @dmitryax @pmcollins
```
Comment (dmitryax) [inline .github/CODEOWNERS:116]: Should it be?

```suggestion
internal/configschema/                                   @open-telemetry/collector-contrib-approvers @mx-psi @dmitryax @atoulme
```

File: .chloggen/internal_configschema.yaml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+note: generate configuration with mdatagen
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [30187]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext: |
+  Components can now generate their own configuration using the following instructions in metadata.yaml:
```
Comment (dmitryax) [inline .chloggen/internal_configschema.yaml:19]:
```suggestion
  Components can now generate their own configuration documentation using the following instructions in metadata.yaml:
```

File: .chloggen/internal_configschema.yaml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  ```
+  config:
+    generate: true
```
Comment (dmitryax) [inline .chloggen/internal_configschema.yaml:22]: Please update this as well

File: receiver/apachereceiver/apache.yaml
Comment (dmitryax) [inline receiver/apachereceiver/apache.yaml:1]: Can we call the file `config_schema.yaml`?

Comment (atoulme) [reply receiver/apachereceiver/apache.yaml:1]: done

**Comment by atoulme** [comment]
Rebased. Please take a look

File: .github/CODEOWNERS (hunk 4 above)
Comment (dmitryax) [reply .github/CODEOWNERS:116]: still the same :)

Comment (atoulme) [reply .github/CODEOWNERS]: Fixed, sorry.

File: receiver/apachereceiver/internal/configgen/component_config_gen.go [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+	"os"
+	"path/filepath"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/configschema"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/apachereceiver"
+)
+
+//go:generate go run ./component_config_gen.go
+
+func main() {
```
Comment (dmitryax) [inline receiver/apachereceiver/internal/configgen/component_config_gen.go:15]: Oh, we bring the file generation to the application runtime... I'm not sure it's a good approach to have documentation generation affect the application runtime. It might delay the collector start or break the collector on some platforms. We don't test mdatagen extensively on all platforms. I believe all the file generation should stay in specific commands or make targets, not the application itself.

I see why you are doing it, and I don't have any similar alternative suggestion, unfortunately.

If we cannot generate the files as part of `make generate` to remove the dependency on all modules from one place, I think we should avoid going through this step and go straight with metadata -> (config structs + docs) generation as was suggested in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/27003.

Comment (dmitryax) [reply receiver/apachereceiver/internal/configgen/component_config_gen.go:15]: Sorry for not bringing this earlier

Comment (dmitryax) [reply receiver/apachereceiver/internal/configgen/component_config_gen.go:15]: Actually, nevermind what I wrote. I confused it with `init()` statement, lol. My head is blurry this morning

File: receiver/apachereceiver/internal/configgen/component_config_gen.go [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+import (
+	"os"
+	"path/filepath"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/configschema"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/apachereceiver"
+)
+
+//go:generate go run ./component_config_gen.go
```
Comment (dmitryax) [inline receiver/apachereceiver/internal/configgen/component_config_gen.go:13]: I'm not sure if go can generate from a generated file in one shot. Likely, it means we have to run `make generate` twice. Maybe we make calling this file explicitly from `make generate` once it's generated instead of putting `go:generate` pragma here?

Comment (atoulme) [reply receiver/apachereceiver/internal/configgen/component_config_gen.go:13]: No, it's recursive. That's the surprising thing I learned through this. The first generation adds the code to generate the docs, and then this code executes, all in one `make generate`.

Comment (atoulme) [reply receiver/apachereceiver/internal/configgen/component_config_gen.go:13]: OK, for the record, I was wrong :/ I will look at this a bit more.

**Comment by atoulme** [comment]
This PR will never be merged since mdatagen has moved over to core. Closing.
